tools/kprint/addBrackets.cpp

```diff
@@ -0,0 +1,359 @@
+  if (data.leftAssoc.count(outerName) && data.leftAssoc.at(outerName).count(innerName) && terminalPos == terminals.size() - 1) {
+    return true;
+  }
+  if (data.rightAssoc.count(outerName) && data.rightAssoc.at(outerName).count(innerName) && terminalPos == 0) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Computes whether a bracket needs to appear around `inner`.
+ *
+ * outer: the parent of `inner`
+ *
+ * leftCapture: the left capture of `inner`
+ *
+ * rightCapture: the right capture of `inner`
+ *
+ * inner: the pattern to test
+ *
+ * position: the position of `inner` within the nonterminals of `outer`
+ *
+ * data: pretty printing metadata
+ *
+ * returns: true if a bracket is required according to this algorithm. Roughly
+ * speaking, parentheses are not needed around associative symbols, constants,
+ * or patterns that appear between two terminals in their parent. Otherwise,
+ * parentheses are needed if priority or associativity forbids the term inside
+ * its parent, or if the term begins or ends with a nonterminal and priority
+ * and associativity do not forbid the term from appearing inside its left
+ * or right capture, respectively, and its left or right capture does not
+ * end or begin, respectively, with a terminal.
+ *
+ * Warning: this algorithm is *unsound*. There is no guarantee that after 
+ * it completes, the resulting term can be unparsed into a string that will
+ * parse unambiguously as the original AST. Indeed, the problem of inserting
+ * brackets into an AST in order to ensure this property is NP-hard, and 
+ * grammars can be constructed such that solving the unparsing problem is 
+ * equivalent to boolean satisfiability. This algorithm is merely a linear-time
+ * approximation of the general problem which performs very well in most
+ * real-world grammars, especially the common expression-grammar category of
+ * grammar where parentheses are most often required in practice.
+ */
+bool requiresBracketWithSimpleAlgorithm(KORECompositePattern *outer, KORECompositePattern *leftCapture, KORECompositePattern *rightCapture, KOREPattern *inner, int position, PrettyPrintData const& data) {
+  if (auto innerComposite = dynamic_cast<KORECompositePattern *>(inner)) {
+    std::string innerName = innerComposite->getConstructor()->getName();
+    if (innerName == outer->getConstructor()->getName()) {
+      if (data.assoc.count(innerName)) {
+        return false;
+      }
+    }
+    if (innerName == "\\dv") {
+      return false;
+    }
+    Fixity fixity = getFixity(position, outer->getConstructor(), data);
+    if (fixity == EMPTY) {
+      return false;
+    }
+    bool priority = isPriorityWrong(outer, innerComposite, position, data);
+    if (priority) {
+      return true;
+    }
+    if (data.terminals.at(innerName) == "0") {
+      return false;
+    }
+    Fixity innerFixity = getFixity(innerComposite->getConstructor(), data);
+    if ((innerFixity & BARE_RIGHT) && rightCapture != nullptr) {
```

Comment:
This if (and the one after, but they are the same thing) was somewhat tricky to understand, so if you'd add an explanation of why it works, it would probably help. Suggestion (the details are probably inaccurate):

"
We are not in one of the cases where we know it's safe to skip adding parentheses.
At this point we know that `outer` does not have higher priority than `inner`, so, if the
inversePriority bool is true, then we must have added parentheses somewhere else
(e.g. around `outer`), so we don't need to add some extra ones here. Otherwise we
add some parentheses around `inner` because  we don't want `inner`'s right
operand to be parsed as a direct child of `rightCapture`.
"

Reply:
I mean, I have no idea how this algorithm works, I just know it does. Somewhere out there on the internet is a master's thesis done by someone in Europe whose algorithm eventually became part of the Maude implementation, from which we took it (and believe me, it was wholly impenetrable to understand in that form). But without access to that paper, I couldn't begin to explain why this particular incantation is the one we use here.

Reply:
Ah, ok. Well, I tried to understand it, so that's why I asked most of these questions...